javalin/src/main/java/io/javalin/http/Context.kt

```diff
@@ -26,132 +23,91 @@ import java.nio.charset.Charset
-    @get:JvmSynthetic @set:JvmSynthetic internal var handlerType = HandlerType.BEFORE
-    @get:JvmSynthetic @set:JvmSynthetic internal var resultReference = AtomicReference<Result<out Any?>>(Result())
-    // @formatter:on
-
-    private val characterEncoding by lazy { ContextUtil.getRequestCharset(this) ?: "UTF-8" }
-
-    private val body by lazy {
-        this.throwContentTooLargeIfContentTooLarge()
-        req.inputStream.readBytes()
-    }
+interface Context {
 
-    /** Gets the handler type of the current handler */
-    fun handlerType(): HandlerType = handlerType
+    /** Servlet request */
+    val req: HttpServletRequest
+    /** Servlet response */
+    val res: HttpServletResponse
 
     /** Gets an attribute from the Javalin instance serving the request */
-    fun <T> appAttribute(key: String): T = appAttributes[key] as T
-
-    fun cookieStore() = cookieStore
-    private val cookieStore by lazy { CookieStore(this) }
-
-    /**
-     * Gets the path that was used to match request (also includes before/after paths)
-     */
-    fun matchedPath() = matchedPath
-
-    /**
-     * Gets the endpoint path that was used to match request (null in before, available in endpoint/after)
-     */
-    fun endpointHandlerPath() = if (handlerType != HandlerType.BEFORE) {
-        endpointHandlerPath
-    } else {
-        throw IllegalStateException("Cannot access the endpoint handler path in a 'BEFORE' handler")
-    }
+    fun <T> appAttribute(key: String): T
+    /** Gets the handler type of the current handler */
+    fun handlerType(): HandlerType
+    /** * Gets the path that was used to match request (also includes before/after paths) */
+    fun matchedPath(): String
+    /** Gets the endpoint path that was used to match request (null in before, available in endpoint/after) */
+    fun endpointHandlerPath(): String
 
     ///////////////////////////////////////////////////////////////
     // Request-ish methods
     ///////////////////////////////////////////////////////////////
 
-    /** Gets the request body as a [String]. */
-    fun body(): String = bodyAsBytes().toString(Charset.forName(characterEncoding))
+    /** Gets the request content length. */
+    fun contentLength(): Int = req.contentLength
+    /** Gets the request content type, or null. */
+    fun contentType(): String? = req.contentType
+    /** Gets the request method. */
+    fun method(): HandlerType
+    /** Gets the request path. */
+    fun path(): String = req.requestURI
+    /** Gets the request port. */
+    fun port(): Int = req.serverPort
+    /** Gets the request protocol. */
+    fun protocol(): String = req.protocol
+    /** Gets the request context path. */
+    fun contextPath(): String = req.contextPath
+    /** Gets the request user agent, or null. */
+    fun userAgent(): String? = req.getHeader(Header.USER_AGENT)
+    /** Try to obtain request encoding from [Header.CONTENT_TYPE] header */
+    fun characterEncoding(): String?
```

Comment:
If this doc is correct, why doesn't it have a default implementation in `req.getHeader(Header.CONTENT_TYPE)` ?

Reply:
It's cached property and I didn't want to implement those methods on both sides to reduce confusion: 

https://github.com/javalin/javalin/blob/9a21d54897d16c2393b9b8e4bd74c828d39e3ba8/javalin/src/main/java/io/javalin/http/ServletContext.kt#L42

Reply:
Can you see why it's cached? It seems very strange 🤔

Reply:
https://github.com/javalin/javalin/commit/5cc44b264b5877fd3069c30daa7e8ad4d553e706

Reply:
I think we can just not cache this, I don't think it deserves special treatment.